protocols/relay/src/v2/client/handler.rs

```diff
@@ -0,0 +1,717 @@
+    Renewing {
+        /// Buffer of messages to be send to the transport listener.
+        pending_msgs: VecDeque<transport::ToListenerMsg>,
+    },
+    None,
+}
+
+impl Reservation {
+    fn accepted(
+        &mut self,
+        renewal_timeout: Option<Delay>,
+        addrs: Vec<Multiaddr>,
+        to_listener: mpsc::Sender<transport::ToListenerMsg>,
+        local_peer_id: PeerId,
+    ) -> Event {
+        let (renewal, mut pending_msgs) = match std::mem::replace(self, Self::None) {
+            Reservation::Accepted { pending_msgs, .. }
+            | Reservation::Renewing { pending_msgs, .. } => (true, pending_msgs),
+            Reservation::None => (false, VecDeque::new()),
+        };
+
+        pending_msgs.push_back(transport::ToListenerMsg::Reservation(Ok(
+            transport::Reservation {
+                addrs: addrs
+                    .into_iter()
+                    .map(|a| {
+                        a.with(Protocol::P2pCircuit)
+                            .with(Protocol::P2p(local_peer_id.into()))
+                    })
+                    .collect(),
+            },
+        )));
+
+        *self = Reservation::Accepted {
+            renewal_timeout,
+            pending_msgs,
+            to_listener,
+        };
+
+        Event::ReservationReqAccepted { renewal }
+    }
+
+    fn failed(&mut self) -> Event {
+        let renewal = matches!(
+            self,
+            Reservation::Accepted { .. } | Reservation::Renewing { .. }
+        );
+
+        *self = Reservation::None;
+
+        Event::ReservationReqFailed { renewal }
+    }
+
+    fn forward_messages_to_transport_listener(&mut self, cx: &mut Context<'_>) {
+        if let Reservation::Accepted {
+            pending_msgs,
+            to_listener,
+            ..
+        } = self
+        {
+            if !pending_msgs.is_empty() {
+                match to_listener.poll_ready(cx) {
+                    Poll::Ready(Ok(())) => {
+                        if let Err(e) = to_listener
+                            .start_send(pending_msgs.pop_front().expect("Called !is_empty()."))
+                        {
+                            debug!("Failed to sent pending message to listener: {:?}", e);
+                            *self = Reservation::None;
+                        }
+                    }
+                    Poll::Ready(Err(e)) => {
+                        debug!("Channel to listener failed: {:?}", e);
+                        *self = Reservation::None;
+                    }
+                    Poll::Pending => {}
+                }
+            }
+        }
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<SubstreamProtocol<outbound_hop::Upgrade, OutboundOpenInfo>>> {
+        self.forward_messages_to_transport_listener(cx);
+
+        let (mut renewal_timeout, pending_msgs, to_listener) =
+            match std::mem::replace(self, Reservation::None) {
```

Comment:
Thoughts on this refactor? https://github.com/MarcoPolo/rust-libp2p/blob/marco/patch1/protocols/relay/src/v2/client/handler.rs#L660.

I have a preference for it because you only assign `*self` once, so I think there's smaller room for error in forgetting to properly assign self. And, more subjectively, I find the flow easier to follow in the given patch.

cc @thomaseizinger

Reply:
It is a bit strange to match on `self` twice. I find the top-level `mem::replace` easier to understand but don't feel strongly about it :)

Reply:
> It is a bit strange to match on self twice.
Agreed. Here's another take on the same idea that I think is a little cleaner:
https://github.com/MarcoPolo/rust-libp2p/blob/702254d9/protocols/relay/src/v2/client/handler.rs#L660

Still only assign *self once.

Reply:
In favor of both:

- Reducing the number of `*self` assignments to reduce room for error.
- Reducing `match` on `self` and thus reducing number of needed `unreachable!` calls.

Unfortunately both are reoccurring issues across the rust-libp2p code base. I haven't seen a viable solution yet.

- Taking `self` instead of `&mut self` pushes the re-assignment to the layer above.
- Crates like https://crates.io/crates/replace_with are [risky in case of panic](https://users.rust-lang.org/t/state-machine-ownership-awkwardness/26693/7).
- ...

dbfe25bf5872700a281d058e62b4adf822f10d56 applies a slightly adjusted version of Marco's latest suggestion. What do you think @thomaseizinger and @MarcoPolo?

Reply:
That change lgtm